pushsnapscript/snap_store.py

```diff
@@ -2,48 +2,220 @@
+        log.warning(e)
+        return
+
+    release_kwargs = {
+        'snap_name': _SNAP_NAME_ON_STORE,
+        'revision': current_snap_revision,
+        'channels': [channel]
+    }
+
+    log.debug('Calling store.release() with these kwargs: {}'.format(snap_file_path))
+    store.release(**release_kwargs)
+
+
+def _list_all_revisions(store):
+    return store.get_snap_revisions(_SNAP_NAME_ON_STORE)
+
+
+def _pluck_metadata(snap_revisions):
+    return {
+        revision['revision']: {
+            'version': GeckoSnapVersion.parse(revision['version']),
+            'current_channels': revision.get('current_channels', []),
+        }
+        for revision in snap_revisions
+    }
+
+
+def _filter_versions_that_are_not_the_same_type(metadata_per_revision, channel):
+    expected_version_type = _VERSION_TYPE_PER_CHANNEL[channel]
+
+    return {
+        revision: revision_metadata
+        for revision, revision_metadata in metadata_per_revision.items()
+        if revision_metadata['version'].version_type == expected_version_type
+    }
+
+
+def _populate_sha3_384(store, metadata_per_revision):
+    return {
+        revision: {
+            **revision_metadata,
+            'download_sha3_384': _get_from_sha3_384_from_revision(store, revision),
+        }
+        for revision, revision_metadata in metadata_per_revision.items()
+        if revision > 1    # First revision doesn't have sha3_384
+    }
+
+
+def _get_from_sha3_384_from_revision(store, revision):
+    # Sadly, this function is not exposed in snapcraft.
+    headers = store.cpi.get_default_headers()
+    headers.update({
+        'Accept': 'application/hal+json',
+        'X-Ubuntu-Series': DEFAULT_SERIES
+    })
+    params = {
+        'fields': 'status,download_sha3_384,revision',
+        'revision': revision
+    }
+    url = 'api/v1/snaps/details/{}'.format(_SNAP_NAME_ON_STORE)
+    resp = store.cpi.get(url, headers=headers, params=params)
+    return resp.json()['download_sha3_384']
+
+
+def _find_revision_and_version_of_current_snap(metadata_per_revision, current_sha3_384):
+    item = get_single_item_from_sequence(
+        metadata_per_revision.items(),
+        lambda item: item[1]['download_sha3_384'] == current_sha3_384,
+        ErrorClass=ValueError,
+        no_item_error_message='No revision has sha3_384 "{}"'.format(current_sha3_384),
+        too_many_item_error_message='Too many revisions have sha3_384 "{}"'.format(current_sha3_384),
+    )
+    revision = item[0]
+    version = item[1]['version']
+    log.debug('Current snap (version "{}") found on the store at revision {}'.format(version, revision))
+    return revision, version
+
+
+def _check_current_snap_is_not_released(current_revision, current_version, latest_released_revision, latest_released_version):
+    if latest_released_revision == current_revision:
+        if latest_released_version == current_version:
+            raise AlreadyLatestError(latest_released_version, latest_released_revision)
```

Comment:
Does the `_find_revision_and_version_of_current_snap` verify that the `sha3_384` of the same-version snap in the store matches the one we're trying to upload? If so, awesome.
If not, someone might be able to upload a malicious, current-version snap before we do, and I believe our snap task will skip the upload and go green. In that case, we may want to either verify the sha from the metadata, or download the snap and verify its sha.

Reply:
Yes, thanks to this condition https://github.com/mozilla-releng/pushsnapscript/pull/14/files#diff-62d14b5f427f0c76b5db0e636464a08cR170 . In it, `current_sha3_384` comes from the local file and `item[1]['download_sha3_384']` from the store. Therefore, if somebody uploaded a malicious snap, then `_find_revision_and_version_of_current_snap` would not return it. It will only return ours.

Anyway, your comment made me realize I haven't considered the scenario where someone publishes a same-version snap with a different sha. In the current implementation, we won't error out. Instead we'll put our version on top of theirs. We need to warn releaseduty is something like this comes up.

Thank you for calling this out!

Reply:
Edge scenario prevented!